src/c1_state_machine/p5_digital_cash.rs

```diff
@@ -48,9 +108,430 @@ impl StateMachine for DigitalCashSystem {
     type State = State;
     type Transition = CashTransaction;
 
-    fn next_state(starting_state: &State, t: &CashTransaction) -> State {
-        todo!("Exercise 1")
+    fn next_state(starting_state: &Self::State, t: &Self::Transition) -> Self::State {
+        match t {
+            CashTransaction::Mint { minter, amount} => handle_mint(starting_state.clone(), minter.clone(), *amount),
+            CashTransaction::Transfer { spends, receives } => handle_transfer(starting_state.clone(), spends.clone(), receives.clone())
+        }
+    }
+}
+
+fn handle_mint(starting_state: State, minter: User, amount: u64) -> State {
+    let mut next_state = starting_state;
+    let bill = Bill { owner: minter.clone(), amount: amount, serial: next_state.next_serial() };
+    next_state.insert_bill(bill);
+    next_state.increment_serial();
+    next_state
+}
+
+fn handle_transfer(starting_state: State, spends: Vec<Bill>, receives: Vec<Bill>) -> State {
+    if spends.is_empty() || receives.is_empty() {
+        return starting_state;
+    }
+    let mut next_state = starting_state.clone();
+
+    {
+        let input_set: BTreeSet<_> = spends.iter().collect();
+        if input_set.len() < spends.len() {
+            return starting_state;
+        }
+    }
+
+    {
+        let output_set: BTreeSet<_> = receives.iter().collect();
+        if output_set.len() < receives.len() {
+            return starting_state;
+        }
+    }
+
+    let mut total_input: u64 = 0;
+    let mut total_output: u64 = 0;
+
+    for input in spends.iter() {
+        let amount = input.amount();
+        let bill = input.clone();
+        if !next_state.contains_bill(bill) {
+            return starting_state;
+        }
+
+        if total_input.checked_add(amount).is_none() {
+            return starting_state;
+        }
+        total_input += amount;
+    }
+
+    for output in receives.iter() {
+        let amount = output.amount;
+        let bill = output.clone();
+        let serial = bill.serial;
+
+        if amount <= 0 {
+            return starting_state;
+        }
+
+        if next_state.next_serial().checked_add(1).is_none() {
+            return starting_state;
+        }
+
+        if next_state.contains_bill(bill) {
+            return starting_state;
+        }
+
+        if total_output.checked_add(amount).is_none() {
+            return starting_state;
+        }
+        total_output += amount;
+    }
+
+    if total_output > total_input {
+        return starting_state;
+    }
+
+    // update storage
+
+    for input in spends {
+        next_state.remove_bill(input);
+    }
+
+    for output in receives {
+        // add bill to the map and increment serial safely.
+        if next_state.next_serial().checked_add(1).is_none() {
+            return starting_state;
+        }
+
+        if output.serial != next_state.next_serial {
+            return starting_state;
+        }
```

Comment:
This made me realize that I didn't design the problem in the nicest way for the user. It would be nice if we didn't require the user to provide the next serial number but rather we just made them provide the owner and amount, then fill the correct serial here. Maybe too big of a change at this point, but figured I'd share the thought.

Reply:
I am going to say for next iteration we can make it simpler. I just dont have time to continue on this have to move on to the other stuff. I am almost finished with the p2p lec now.